### PR TITLE
RPS Challenge

### PR DESCRIPTION
**Pull request by christopherxwebb**
I'm checking in a partial solution. So far, I have not been able to implement the logic that decides the outcome of the game.
I am unsure about how to make the different elements of the MVC speak to each. Some work ok: it's the final page that is the problem. Any thoughts or feedback would be greatly appreciated.

### CODE REVIEW

File: app/rps.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    post '/weapons' do
+        $player = params[:player_1]
+        $computer = Computer.new.compmove
+        erb :weapons
+    end
+    
+    post '/result' do
+        $weapon = params[:p1move]
+        $computer = Computer.new.compmove
+       # $outcome = Game.new.outcome #This seems to be the problem when I run sinatra.
```
Comment (sherwingp) [inline app/rps.rb]: You forgot to require game.rb

File: lib/game.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
+require_relative 'computer'
+require './app/rps'
+
+class Game
+    attr_reader :outcome
+
+def compmove
```
Comment (sherwingp) [inline lib/game.rb]: I can see what your intention is, but every time you call the method compmove, it is selecting a random choice again because of the .sample and assigning it to @compmove. You should move @compmove to your initialise method and add an attribute reader for it

File: app/rps.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    post '/weapons' do
+        $player = params[:player_1]
+        $computer = Computer.new.compmove
+        erb :weapons
+    end
+    
+    post '/result' do
+        $weapon = params[:p1move]
+        $computer = Computer.new.compmove
```
Comment (sherwingp) [inline app/rps.rb]: This is probably a source of a bug. You are correctly using p1move, that is the weapon you chose. But you are reassigning computer to a **new** instance of Computer. You don't have to do this because you already assigned an instance of it to the global variable $computer in the /weapons route

File: lib/computer.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+class Computer
+   attr_reader :compmove
+    def compmove
```
Comment (sherwingp) [inline lib/computer.rb:3]: I noticed this method is also present in game.rb. One of these should be removed

File: app/rps.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+       # $outcome = Game.new.outcome #This seems to be the problem when I run sinatra.
+        erb :result
+    end
+
+    # post '/game' do
```
Comment (sherwingp) [inline app/rps.rb]: Yes, I think these routes are redundant now.

File: app/views/game.erb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<h1> You chose <%=@weapon%> </h1>
```
Comment (sherwingp) [inline app/views/game.erb:1]: This view is unused

File: app/views/result.erb [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<h1> You chose <%= $weapon %>.<h1>
+<h1> Your opponent chose <%= $computer %>.</h1>
+
+<%# # I want to write: 'You lose/win/draw' by using the global variable of $game. %>
```
Comment (sherwingp) [inline app/views/result.erb]: Remember to add this implementation once you fix the underlying bug in your model (Game.rb)

File: app/rps.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+      register Sinatra::Reloader
+    end
+
+    get '/' do
+        erb :index
+    end
+
+    post '/weapons' do
+        $player = params[:player_1]
+        $computer = Computer.new.compmove
```
Comment (sherwingp) [inline app/rps.rb]: We assigned the computer move to $computer here. But once you call the outcome method in the /results route, is it using the computer move assigned here?

File: README.md [hunk 9]
```diff
@@ -13,7 +15,7 @@ Task
 
 Knowing how to build web applications is getting us almost there as web developers!
 
-The Makers Academy Marketing Array ( **MAMA** ) have asked us to provide a game for them. Their daily grind is pretty tough and they need time to steam a little.
+The Makers Academy Marketing Array ( **MAMA** ) have asked us to provide a game for them. Their daily grind is pretty tough and they need time to let off steam a little.
```
Comment (alicelieutier) [inline README.md:18]: If you make a separate PR with just this change, I'll merge it to the course repo ;)

File: app/rps.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
+require "sinatra/base"
+require "sinatra/reloader"
+require "./lib/player.rb"
+require "./lib/computer.rb"
+require "./lib/game.rb"
+
+class RPS < Sinatra::Base
+    enable :sessions
```
Comment (alicelieutier) [inline app/rps.rb:8]: You enable sessions but do not use them afterwards, instead preferring a global variable.
I actually think the code would be nicer using sessions.

File: app/views/game.erb (hunk 6 above)
Comment (alicelieutier) [reply app/views/game.erb:1]: @christopherxwebb you marked this as resolved, but the view is still unused. Maybe delete this file?

File: app/rps.rb [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+        $player = params[:player_1]
+        $cpu = Computer.new.compmove
+        erb :weapons
+    end
+    
+    post '/result' do
+        $weapon = params[:p1move]
+        $outcome = Game.new.result
+        erb :result
+    end
```
Comment (alicelieutier) [inline app/rps.rb:27]: You may want to follow the post -> redirect -> get pattern here instead or returning views from the post routes. Check https://github.com/makersacademy/course/blob/main/intro_to_the_web/post_redirect_get_pattern.md

File: lib/game.rb [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+def result
+  if $weapon == "rock" && $cpu == "scissors" || $weapon == 'paper' && $cpu == "rock" || $weapon == 'scissors' && $cpu == "paper"
+    :win
+  elsif $weapon == $cpu
+    :draw
+  else
+    :lose
+    end
+  end
```
Comment (alicelieutier) [inline lib/game.rb:16]: You are accessing global variables from this class. This is not ideal. Basically, your model only works because it relies on the controller setting all these global variables in the right way, and you classes don't speak with eachother.
When you develop a web app, the model part of it should still very much work on its own, like an irb app.
Then the controller and views are just there to make the interface better for the user.

One way you can do this is developing the model as an irb app, using feature tests (in irb or rspec) and unit tests.

In anycase, you should be passing the player (or player move) and the computer (or computer move) to the game class through arguments, as opposed to the game class going and reading global variables outside itself.

File: lib/player.rb [hunk 13]
```diff
@@ -0,0 +1,3 @@
+# class Player
```
Comment (alicelieutier) [inline lib/player.rb:1]: Delete unused files.

File: spec/features/enter_name_spec.rb [hunk 14]
```diff
@@ -0,0 +1,8 @@
+feature 'Enter names' do
+    scenario 'submitting names' do
+        visit ('/')
+        fill_in :player_1, with: 'Player'
+        click_button 'Submit'
+        expect(page).to have_content 'Ok Player choose your weapon:'
+    end
+end
```
Comment (alicelieutier) [inline spec/features/enter_name_spec.rb:8]: Good test! It would be nice to see a second one, to check that the screen does adapt to the name being entered, and doesn't always reply `'Ok Player choose your weapon:'` even if I enter 'Alice' for example.
You could also test an error scenario, for example there could be an error if the user leaves the player field empty.

File: spec/features/pick_weapon_spec.rb [hunk 15]
```diff
@@ -0,0 +1,9 @@
+feature 'Picking Weapon' do
+    scenario 'to choose weapon against CPU' do
+        visit ('/')
+        fill_in :player_1, with: 'Player'
+        click_button 'Submit'
+        click_button 'rock'
+        expect(page).to have_content 'You chose rock'
+    end
+end
```
Comment (alicelieutier) [inline spec/features/pick_weapon_spec.rb:9]: Same comment as before. This is a good test, but you should add more tests for each feature.

Comment (alicelieutier) [reply spec/features/pick_weapon_spec.rb:9]: Also, by stubbing randomness (check `srand`) you could test the full game, including the result page.

**Review by alicelieutier — COMMENTED** [review]
It would be nice to see some unit tests for the game class, as it does some heavy logic with calculating the results.
You class organisation in the model can be improved to function more as an independant part, and be less coupled to your controller. See my comments below.

File: README.md (hunk 9 above)
Comment (christopherxwebb) [reply README.md:18]: haha thanks Alice. I was being pedantic here!

File: app/rps.rb (hunk 10 above)
Comment (christopherxwebb) [reply app/rps.rb:8]: Ok great. This has helped because I was unsure about sessions, what they did. I understand they temporarily store form data but need to do a bit more research on how they actually work.

File: app/rps.rb (hunk 11 above)
Comment (christopherxwebb) [reply app/rps.rb:27]: Great, thanks Alice, I'll take a look now!

File: lib/player.rb (hunk 13 above)
Comment (christopherxwebb) [reply lib/player.rb:1]: Really bad at (not) doing this. I always like to keep things open but as you rightly suggest makes things way more complicated.

File: spec/features/enter_name_spec.rb (hunk 14 above)
Comment (christopherxwebb) [reply spec/features/enter_name_spec.rb:8]: yes great idea!

File: spec/features/pick_weapon_spec.rb (hunk 15 above)
Comment (christopherxwebb) [reply spec/features/pick_weapon_spec.rb:9]: Ah, yes, good point! I'll look into stubbing later on today. Thank you for all these Alice, really helpful!
